js/src/path.js

```diff
@@ -4,43 +4,220 @@
-    });
+      let part: Part;
+      if (idx != null) {
+        part = new IndexPath(idx, intoKey);
+      } else if (h != null) {
+        part = new HashIndexPath(h, intoKey);
+      } else {
+        throw new Error('unreachable');
+      }
+      this._parts.push(part);
+      this._parse(rem);
+      return;
+    }
+
+    if (op === ']') {
+      throw new SyntaxError('] is missing opening [');
+    }
+
+    throw new SyntaxError(`Invalid operator: ${op}`);
+  }
+
+  _parsePathIndex(str: string): [?indexType, ?Hash, string] {
+    if (str[0] === '"') {
+      // String is complicated because ] might be quoted, and " or \ might be escaped.
+      const stringBuf = [];
+      let i = 1;
+
+      for (; i < str.length; i++) {
+        let c = str[i];
+        if (c === '"') {
+          break;
+        }
+        if (c === '\\' && i < str.length - 1) {
+          i++;
+          c = str[i];
+          if (c !== '\\' && c !== '"') {
+            throw new SyntaxError('Only " and \\ can be escaped');
+          }
+        }
+        stringBuf.push(c);
+      }
+
+      if (i === str.length) {
+        throw new SyntaxError('[ is missing closing ]');
+      }
+      return [stringBuf.join(''), null, str.slice(i + 2)];
+    }
+
+    const closingIdx = str.indexOf(']');
+    if (closingIdx === -1) {
+      throw new SyntaxError('[ is missing closing ]');
+    }
+
+    const idxStr = str.slice(0, closingIdx);
+    const rem = str.slice(closingIdx + 1);
+
+    if (idxStr.length === 0) {
+      throw new SyntaxError('Empty index value');
+    }
+
+    if (idxStr[0] === '#') {
+      const hashStr = idxStr.slice(1);
+      const h = Hash.parse(hashStr);
+      if (h == null) {
+        throw new SyntaxError(`Invalid hash: ${hashStr}`);
+      }
+      return [null, h, rem];
+    }
+
+    if (idxStr === 'true') {
+      return [true, null, rem];
+    }
+
+    if (idxStr === 'false') {
+      return [false, null, rem];
+    }
+
+    const n = Number(idxStr);
+    if (!Number.isNaN(n)) {
+      return [n, null, rem];
+    }
+
+    throw new SyntaxError(`Invalid index: ${idxStr}`);
+  }
+
+  _getAnnotation(str: string): [string /* ann */, string /* rem */] {
+    const parts = annotationRe.exec(str);
+    if (parts) {
+      invariant(parts.length === 2);
+      return [parts[1], str.slice(parts[0].length)];
+    }
+    return ['', str];
+  }
+}
+
+/**
+ * Gets Struct field values by name.
+ */
+export class FieldPath {
+  /**
+   * The name of the field, e.g. `.Name`.
+   */
+  name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+
+  resolve(value: Value): Promise<?Value> {
+    const t = getTypeOfValue(value);
+    if (t.kind !== Kind.Struct) {
+      return Promise.resolve();
+    }
+
+    const f = (t.desc: StructDesc).getField(this.name);
+    if (!f) {
+      return Promise.resolve(); // non-present field
+    }
+
+    // $FlowIssue: Flow doesn't know that it's safe to just access the field name here.
```

Comment:
I wonder if we could do `(value: {[key: string]: ?Value})[this.name]`... not sure if it is better though.